### PR TITLE
Add unsupported operation visitor as fallback for codegen

### LINKED ISSUE #327: Implement unsupported dialect
Since VAST cannot model every AST (and it will take a while since it will be able) provide a default unsupported dialect.

If we want to be fancy we can provide more unsupported operations, but for now let's begin with a generic operation in ~following shape:

```
%A = vast.unsupported< "some string data to provide more visibility into what failed" >
{
   properly lowered child nodes if any present
}
```

### PR DESCRIPTION
**Pull request by kumarak**
Resolves #327

- Implement unsupported operation dialects
- Add Unsupported Operation visitor as another fallback layer in codegen
- Add support for lowering the unsupported AST node using unsupported op dialects

### CODE REVIEW

File: include/vast/Translation/CodeGenContext.hpp [hunk 1]
```diff
@@ -161,6 +161,9 @@ namespace vast::cg
                 if (llvm::isa< clang::FunctionDecl >(dctx))
                     continue;
 
+                if (llvm::isa< clang::LinkageSpecDecl >(dctx))
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenContext.hpp:164]: How is this related to the unsupported dialect? I mean having some extra small-ish change(s) thrown in a bigger PR is fine, but the commit should be self-contained and ideally mention it in the PR description.

File: include/vast/Dialect/HighLevel/HighLevelUnsupported.td [hunk 2]
```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2022-present, Trail of Bits, Inc.
+
+#ifndef VAST_DIALECT_HIGHLEVEL_IR_HIGHLEVELUNSUPPORTED
+#define VAST_DIALECT_HIGHLEVEL_IR_HIGHLEVELUNSUPPORTED
+
+def HighLevel_UnsupportedDecl
```
Comment (lkorenc) [inline include/vast/Dialect/HighLevel/HighLevelUnsupported.td:6]: We want to have a separate dialect for unsupported operations.

File: include/vast/Translation/CodeGen.hpp [hunk 3]
```diff
@@ -579,7 +579,7 @@ namespace vast::cg
         }
 
         operation visit_var_decl(const clang::VarDecl *decl) {
-            return _visitor->VisitVarDecl(decl);
+            return _visitor->Visit(decl);
```
Comment (lkorenc) [inline include/vast/Translation/CodeGen.hpp:582]: Separate commit.

File: include/vast/Translation/CodeGenFallBackVisitor.hpp [hunk 4]
```diff
@@ -31,21 +32,54 @@ namespace vast::cg
...
         using DeclVisitor = DefaultFallBackDeclVisitorMixin< Derived >;
         using StmtVisitor = DefaultFallBackStmtVisitorMixin< Derived >;
         using TypeVisitor = DefaultFallBackTypeVisitorMixin< Derived >;
 
-        using DeclVisitor::Visit;
-        using StmtVisitor::Visit;
-        using TypeVisitor::Visit;
+        using UnsupportedVisitor = FallBackMixin< Derived >;
+
+        Operation* Visit(const clang::Stmt *stmt) {
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenFallBackVisitor.hpp:50]: I am really confused why we are doing dispatch here again, when a few lines below it is done for us? `CodeGenFallBackVisitorMixin`.

Comment (lkorenc) [reply include/vast/Translation/CodeGenFallBackVisitor.hpp:50]: Shouldn't the only change to this class be change of those type alias of partial Visitors to the `Unsupported` versions?

File: include/vast/Translation/CodeGenFallBackVisitor.hpp [hunk 5]
```diff
@@ -31,21 +32,54 @@ namespace vast::cg
...
+        Type Visit(clang::QualType type) {
+            if (auto result = UnsupportedVisitor::Visit(type))
+                return result;
+
+            return TypeVisitor::Visit(type);
+        }
     };
 
+    template< typename Derived >
+    using DefaultFallbackVisitorConfig = DefaultFallBackVisitorMixin< Derived,
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenFallBackVisitor.hpp:76]: Keep the default for now as it was, instead provide extra helpers for unsupported one, so you can easily plug it into the whole stack (I think not trying to use it as a default for now will help flesh out the implementation).

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+#include "vast/Translation/CodeGen.hpp"
+#include "vast/Translation/CodeGenContext.hpp"
+#include "vast/Translation/CodeGenVisitor.hpp"
+#include "vast/Translation/CodeGenBuilder.hpp"
+#include "vast/Translation/CodeGenFallBackVisitor.hpp"
+
+namespace vast::cg
+{
+
+#define VISITOR_DECL(decl_type) \
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: Unscoped name that is not properly unique. Also undefine at the end.

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+
+		using LensType::derived;
+		using LensType::context;
+		using LensType::mcontext;
+		using LensType::acontext;
+
+		using LensType::meta_location;
+		using LensType::visit;
+
+		template< typename Op, typename... Args >
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: I am pretty sure all of these helpers are copied from other parts of `cg::` (or at least they look similar-ish enough to be considered a good candidates for shared utilities) - cannot they be re-used?

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+	{
+	   	using DeclVisitor = UnsupportedDeclVisitor< Derived >;
+	    using StmtVisitor = UnsupportedStmtVisitor< Derived >;
+	    using TypeVisitor = UnsupportedTypeVisitor< Derived >;
+
+	    using DeclVisitor::Visit;
+	    using StmtVisitor::Visit;
+	    using TypeVisitor::Visit;
+
+	    operation Visit(const clang::Stmt *stmt) {
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: Why are these methods here? Except for the `Type` overload, which has custom logic.

File: include/vast/Translation/CodeGenFallBackVisitor.hpp [hunk 9]
```diff
@@ -31,21 +32,54 @@ namespace vast::cg
         }
     };
 
-    template< typename Derived >
+    template< typename Derived
+        , template< typename > typename FallBackMixin = UnsupportedFallBackVisitor
```
Comment (lkorenc) [inline include/vast/Translation/CodeGenFallBackVisitor.hpp:36]: `Unsupported` should not be anywhere in this class -> it should be plugged into the "dispatcher" class below as a template argument only (no extra type aliases or anything should be added to it).

**Review by lkorenc — CHANGES_REQUESTED** [review]

File: include/vast/Translation/CodeGenContext.hpp (hunk 1 above)
Comment (kumarak) [reply include/vast/Translation/CodeGenContext.hpp:164]: If you are walking the decl context parent, you may find `LinkageSpecDecl` as one of the parent and not a `NamedDecl`. That will lead to reaching the unreachable. 

I will move it in the separate commit under this PR to avoid mixup with other changes.

File: include/vast/Dialect/HighLevel/HighLevelUnsupported.td (hunk 2 above)
Comment (kumarak) [reply include/vast/Dialect/HighLevel/HighLevelUnsupported.td:6]: @lkorenc, Should the unsupported operation dialect named differently. I used `HighLevel_` as prefix because it is high level dialect. The dialects are of three types. The `UnsupportedDecl` is used only for lowering the Decl node. `UnsupportedExpr` and `UnsupportedOp` are used for lowering stmt node.

Could you explain what do you mean by saying "want to have a separate dialect for unsupported operations"?

Comment (lkorenc) [reply include/vast/Dialect/HighLevel/HighLevelUnsupported.td:6]: Yeah I do not want it to be part of high level dialect. I would like a separate dialect - unsupported - same way as we have `ll` and `core` etc. alongside the `hl`. And this unsupported dialect will have those 3 operations.

Or do you think there is a reason this cannot be achieved/is not a good idea?

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp (hunk 6 above)
Comment (kumarak) [reply include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: undef the macro at the end to avoid conflict with other macros with same name.

File: include/vast/Dialect/HighLevel/HighLevelUnsupported.td (hunk 2 above)
Comment (kumarak) [reply include/vast/Dialect/HighLevel/HighLevelUnsupported.td:6]: @lkorenc, Thanks for explaining it. We can have `unsupported` as a different set of dialect similar to `core`, or `low-level`. The one concern I have is the coverage scope for these dialects are limited. In that case is it a good idea to have them as separate dialect set? I am not sure.

I will have these changes in a separate commit and we can review how much complexity does it add, if we introduce a new set of dialect for unsupported. We can then decide if we should go ahead with that or not.

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp (hunk 7 above)
Comment (kumarak) [reply include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: Most of these utility functions are custom and written similar to one in `cg::Builder`. These utility function buildup region and generate operation specific to the information we need to present with the unsupported operations. Moving them to the common place in builder will not be a good idea. 

I may refactor them a few and see if they can use already existing utility methods from `cg::Builder`.

File: include/vast/Translation/CodeGenUnsupportedVisitor.hpp (hunk 8 above)
Comment (kumarak) [reply include/vast/Translation/CodeGenUnsupportedVisitor.hpp]: Removed this additional indirection!

File: include/vast/Dialect/Unsupported/Unsupported.td [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+include "vast/Interfaces/SymbolInterface.td"
+
+def Unsupported_Dialect : Dialect {
+    let name = "unsupported";
+    let cppNamespace = "::vast::us";
+
+    let summary = "A vast unsupported dialect.";
+
+    let description = [{
+        This dialect intends capture user metadata
```
Comment (xlauko) [inline include/vast/Dialect/Unsupported/Unsupported.td:19]: copy-paste from meta dialect

File: include/vast/Dialect/Unsupported/UnsupportedTypes.td [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+#define VAST_DIALECT_IR_UNSUPPORTED_TYPES
+
+def UnsupportedType
+    : Unsupported_Type< "Unsupported" > {
+  let mnemonic = "type";
+  let parameters = (ins "Type":$elementType);
+  let assemblyFormat = "`<` $elementType `>`";
+}
+
+def UnsupportedVoidType : Unsupported_Type< "UnsupportedVoid" > {
```
Comment (xlauko) [inline include/vast/Dialect/Unsupported/UnsupportedTypes.td:13]: when do you need such type?

File: include/vast/Translation/CodeGenFallBackVisitor.hpp [hunk 12]
```diff
@@ -56,13 +58,16 @@ namespace vast::cg
     template< typename Derived
         , template< typename > typename DefaultMixin
         , template< typename > typename FallBackMixin = DefaultFallBackVisitorMixin
+        , template< typename > typename UnsupporteFallbackdMixin = UnsupportedFallBackVisitor
```
Comment (xlauko) [inline include/vast/Translation/CodeGenFallBackVisitor.hpp:61]: typo Fallbackd -> Fallback

File: include/vast/Translation/CodeGenFallBackVisitor.hpp [hunk 13]
```diff
@@ -56,13 +58,16 @@ namespace vast::cg
...
     >
     struct CodeGenFallBackVisitorMixin
         : DefaultMixin< Derived >
         , FallBackMixin< Derived >
+        , UnsupporteFallbackdMixin< Derived >
```
Comment (xlauko) [inline include/vast/Translation/CodeGenFallBackVisitor.hpp:66]: This should be parameter not part of this class, ie.e you pass `UnsupporteFallbackdMixin` as `FallBackMixin`.

I can fix this and integrate to codegen with option if you like?

Comment (kumarak) [reply include/vast/Translation/CodeGenFallBackVisitor.hpp:66]: @xlauko, Do you mean creating a config that takes `UnsupportedFallBackVisitor` as one of the parameter. Something like defining the config as:
```
    template< typename Derived >
    using DefaultCodeGenVisitorConfig = CodeGenFallBackVisitorMixin< Derived,
        DefaultCodeGenVisitorMixin,
        UnsupportedFallBackVisitor
    >;
```

One of the reason I didn't add it to `DefaultCodeGenVisitorConfig` because I want both unsupported visitor and default fallback visitor. Previous I updated `DefaultFallBackVisitorMixin` and passed a config as parameter that will multiplex between Unsupported Visitor and Default fallback visitor. It was similar to something like
```
template< typename Derived >
    using DefaultFallbackVisitorConfig = DefaultFallBackVisitorMixin< Derived, UnsupportedFallBackVisitor>;

// pass config to DefaultCodeGenVisitorConfig

    template< typename Derived >
    using DefaultCodeGenVisitorConfig = CodeGenFallBackVisitorMixin< Derived,
        DefaultCodeGenVisitorMixin,
        DefaultFallbackVisitorConfig
    >;
```

This will allow choosing between the default fallback and unsupported visitor in class `DefaultFallBackVisitorMixin` but that mean adding another redirection there.

Can you write a pseudo code of how you like it to be integrated? I will try getting that working else will ask you to update the update it with the options.

Comment (xlauko) [reply include/vast/Translation/CodeGenFallBackVisitor.hpp]: Thats exactly how it was designed to be used. You don't need to choose anything in `DefaultFallBackVisitorMixin`.
FallBack mixin takes care of trying all visitors in order.

Under the hood it will generate the same code as you have now, which essentially cascades visits `default -> unsupported -> default error`.

File: include/vast/Dialect/HighLevel/HighLevelUnsupported.td (hunk 2 above)
Comment (xlauko) [reply include/vast/Dialect/HighLevel/HighLevelUnsupported.td:6]: I think we can remove this now. Unsupported dialect is the way to go, because we can have eventually something that does not use `hl` dialect but still leverage `unsupported` stuff.

File: include/vast/Dialect/Unsupported/UnsupportedTypes.td (hunk 11 above)
Comment (xlauko) [reply include/vast/Dialect/Unsupported/UnsupportedTypes.td:13]: I think it can be merged to single `UnsupportedType` where you make `elementType` optional:

```
  let assemblyFormat = "`<` ( $elementType^ )? `>`";
```

or at least change the name as `UnsupportedVoidType` is a bit confusing.
Use something like `UnsupportedOpaqueType`

File: include/vast/Translation/CodeGenFallBackVisitor.hpp (hunk 13 above)
Comment (kumarak) [reply include/vast/Translation/CodeGenFallBackVisitor.hpp]: @xlauko, I added `FallbackVisitorConfig` to chain the unsupported op fallback. I think this is how it is intended to be chained with `DefaultFallBackVisitorMixin`. Let me know if it looks good.

File: include/vast/Dialect/Unsupported/UnsupportedTypes.td (hunk 11 above)
Comment (kumarak) [reply include/vast/Dialect/Unsupported/UnsupportedTypes.td]: I removed `UnsupportedVoidType` and used the `OptionalParameter` for `elementType`. We can use it for generating types in opaque format as well where we don't visit the underlying types.

**Comment by xlauko** [comment]
Closes #366 #365

**Review by lkorenc — APPROVED** [review]

**Comment by lkorenc** [comment]
Some commits are a mess, I would be glad if they were cleaned up.

**Comment by kumarak** [comment]
@xlauko, Refactoring looks good but I do see it hits llvm unreachable during binding the callback values with the operations. I am seeing with most of my unit test cases whenever it reaches Unsupported visitor.

```
UNREACHABLE executed at /Users/akkumar/Documents/workspace/cxx-common/vcpkg/buildtrees/llvm-16/src/org-16.0.5-90a3d25b70.clean/clang/lib/AST/DeclBase.cpp:946!
```

**Comment by xlauko** [comment]
@kumarak can you provide your tests somewhere please.

**Comment by kumarak** [comment]
Unit test looks good to me now. I see one failure but that is related to lowering c++ default copy constructor #347 and not unsupported visitor.

**Review by lkorenc — APPROVED** [review]

**Review by xlauko — APPROVED** [review]

**Review by xlauko — APPROVED** [review]
